safenode/src/network_transfers/error.rs

```diff
@@ -0,0 +1,95 @@
+// Copyright 2023 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use crate::protocol::{address::DbcAddress, fees};
+
+use sn_dbc::{Error as DbcError, SignedSpend, Token};
+
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeSet;
+use thiserror::Error;
+
+/// The type returned by the `sn_routing` message handling methods.
+pub(crate) type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// Transfer errors.
+#[derive(Error, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub enum Error {
+    #[error("The transfer fee is missing.")]
+    MissingFee,
+    #[error("Invalid fee blinded amount.")]
+    InvalidFeeBlindedAmount,
+    #[error("Too low amount for the transfer fee: {paid}. Min required: {required}.")]
+    FeeTooLow { paid: Token, required: Token },
+    #[error(transparent)]
+    Fees(#[from] fees::Error),
+    #[error("Contacting close group of parent spends failed: {0}.")]
+    SpendParentCloseGroupIssue(String),
+    /// An error from the `sn_dbc` crate.
+    #[error("Dbc error: {0}")]
+    Dbcs(String),
+    /// Spend not found.
+    #[error("Spend not found: {0:?}")]
+    SpendNotFound(DbcAddress),
+    /// A double spend attempt was detected.
+    #[error("A double spend attempt was detected. Incoming and existing spend are not the same: {new:?}. Existing: {existing:?}")]
+    DoubleSpendAttempt {
+        /// New spend that we received.
+        new: Box<SignedSpend>,
+        /// Existing spend of same id that we already have.
+        existing: Box<SignedSpend>,
+    },
+    /// We were notified about a double spend attempt, but they were for different dbcs.
+    #[error("We were notified about a double spend attempt, but they were for different dbcs: {0:?}. Existing: {1:?}")]
+    NotADoubleSpendAttempt(Box<SignedSpend>, Box<SignedSpend>),
+    /// One or more parent spends of a requested spend had a different dst tx hash than the signed spend src tx hash.
+    #[error(
+        "The signed spend src tx ({signed_src_tx_hash:?}) did not match the provided source tx's hash: {provided_src_tx_hash:?}"
+    )]
+    SignedSrcTxHashDoesNotMatchProvidedSrcTxHash {
```

Comment:
I suggest to define a `Custom` error type that message can be formatted.

Reply:
Good that you pointed out that one. I don't like the name of that error.. But it will have to be improved.

What custom error is it you propose? The error we're looking at is a custom error.

Reply:
Custom or Genera error, that can have formatted string.  like
```rust
General(String)
```

That can be used like:
```rust
return Err(Error::General(format!("The signed spend src tx ({signed_src_tx_hash:?}) did not match the provided source tx's hash: {provided_src_tx_hash:?}"))
```

Reply:
Alright, yeah ok I understand.

We've been trying to avoid that actually, and design specific errors for each and every situation requiring an error.

Reply:
That error name I got, with not too much time spent on it, is horrendous though :)

Reply:
I'm testing with a rename to `TxTrailMismatch`, and let the msg for it explain closer.

Reply:
Fixed it in follow up PR.